### PR TITLE
Blocking users by matching email domains suffixes

### LINKED ISSUE #9408: Pattern match email domains for blocking abusers
Currently we're only able to blacklist complete email domains, but seeing abusers using subdomains to circumvent this.
We should be able to pattern match domain suffixes.

### PR DESCRIPTION
**Pull request by geropl**
## Related Issue(s)

Fixes #9408 .

## How to test
- run tests and see they are green
- try to signup [here](https://gpl-9408-edomains.staging.gitpod-dev.com/workspaces) with an primary email that ends on `itpod.io`
- try to open a workspace and see that you are blocked :x: 

## Release Notes

```release-note
block abusers based on email domain suffixes
```

### CODE REVIEW

**Review by jankeromnes — APPROVED** [review]
Super cool, looks good to me and works as advertised. 👍 Many thanks!

Also big kudos on adding tests. :100: 

Please simply fix the two micro-nits, then I'll quickly re-approve. 😁🚀

/hold

**Review by jankeromnes — APPROVED** [review]
Super cool, looks good to me, thanks! 👍

Please simply fix the two micro-nits, then I'll quickly re-approve. 😁🚀

/hold

File: components/gitpod-db/src/typeorm/email-domain-filter-db-impl.ts [hunk 1]
```diff
@@ -29,13 +29,14 @@ export class EmailDomainFilterDBImpl implements EmailDomainFilterDB {
         await repo.save(entry);
     }
 
-    async filter(domain: string): Promise<boolean> {
+    async isBlocked(domain: string): Promise<boolean> {
         const repo = await this.getRepo();
         const result = await repo
             .createQueryBuilder("entry")
-            .where(`entry.domain = :domain`, { domain: domain })
+            .where(`:domain LIKE entry.domain`, { domain: domain })
```
[the remark this answers is not in the document]
Comment (AlexTugarev) [reply components/gitpod-db/src/typeorm/email-domain-filter-db-impl.ts:36]: entry.domain contains the LIKE pattern.

nice solution BTW, reading test cases, this looks good!

[the remark this answers is not in the document]
Comment (jankeromnes) [reply components/gitpod-db/src/typeorm/email-domain-filter-db-impl.ts:36]: Ah, 100% true 👍

File: components/gitpod-db/src/email-domain-filter-db.ts [hunk 2]
```diff
@@ -14,5 +14,5 @@ export interface EmailDomainFilterDB {
      * @param emailDomain
      * @returns false iff this emailDomain is meant to be blocked, aka "filtered out from the list"
```
[the remark this answers is not in the document]
Comment (geropl) [reply components/gitpod-db/src/email-domain-filter-db.ts:15]: Ugh, yeah this should be fixed. :grimacing:

File: components/server/ee/src/auth/email-domain-service.ts [hunk 3]
```diff
@@ -25,7 +25,7 @@ export class EMailDomainServiceImpl implements EMailDomainService {
 
     async isBlocked(email: string): Promise<boolean> {
         const { domain } = this.parseMail(email);
-        return !(await this.domainFilterDb.filter(domain));
+        return this.domainFilterDb.isBlocked(domain);
```
[the remark this answers is not in the document]
Comment (jankeromnes) [reply components/server/ee/src/auth/email-domain-service.ts:28]: Agree, and I also do it quite frequently -- but in this particular case, where we did try to "not" a Promise, I wanted to be particularly cautious and add an `await`. But, I'd also be fine to merge this without the extra safety.

**Review by jankeromnes — APPROVED** [review]
Still looks good to me after the [force-push](https://github.com/gitpod-io/gitpod/compare/79d8d1d992ff4368034141ec207003421fc7be71..39c4ec5b00323a9ce65dec4fc562f91f05fea9ad). 👍

**Pull request merged**